File: leet/0849_maximize_distance_to_closest_person/0849_maximize_distance_to_closest_person.js
// LeetCode 849. Maximize Distance to Closest Person
// https://leetcode.com/problems/maximize-distance-to-closest-person/
// O(n) time, O(1) space
var maxDistToClosest = function(seats) {
    const distanceFromLeft = Array(seats.length).fill(0);
    let lastPerson = -1, maxDistance = 0;
    for (let i = 0; i <= seats.length; i++) {
        if (i == seats.length || seats[i] == 1) {
            let distance;
            if (lastPerson == -1)
                distance = i;
            else if (i == seats.length)
                distance = i - 1 - lastPerson;
            else
                distance = Math.trunc((i - lastPerson) / 2);
            lastPerson = i;
            maxDistance = Math.max(maxDistance, distance);
        }
    }
    return maxDistance;
};

File: leet/0849_maximize_distance_to_closest_person/0849_maximize_distance_to_closest_person.c
// LeetCode 849. Maximize Distance to Closest Person
// https://leetcode.com/problems/maximize-distance-to-closest-person/
// O(n) time-and-space
#define MIN(x,y) ((x) < (y) ? (x) : (y))
#define MAX(x,y) ((x) > (y) ? (x) : (y))
int maxDistToClosest(int* seats, int seatsSize) {
    int *distanceFromLeft = malloc(seatsSize * sizeof(int)),
        *distanceFromRight = malloc(seatsSize * sizeof(int));
    for (int i = 0, j = seatsSize - 1; i < seatsSize; i++, j--) {
        if (seats[i] == 1)
            distanceFromLeft[i] = 0;
        else
            distanceFromLeft[i] = !i || distanceFromLeft[i - 1] == INT_MAX
                ? INT_MAX : distanceFromLeft[i - 1] + 1;
        
        if (seats[j] == 1)
            distanceFromRight[j] = 0;
        else
            distanceFromRight[j] = !i || distanceFromRight[j + 1] == INT_MAX
                ? INT_MAX : distanceFromRight[j + 1] + 1;
    }
    int maxDistance = 0;
    for (int i = 0; i < seatsSize; i++) {
        int distance = MIN(distanceFromRight[i], distanceFromLeft[i]);
        maxDistance = MAX(maxDistance, distance);
    }
    return maxDistance;
}
